roachtest: schemaChangeStep is skipped in acceptance/version-upgrade

[(roachtest).acceptance/version-upgrade failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2561359&tab=buildLog) on [master@cee475331ca3629b503cd2e7c7919b72c98a5ca5](https://github.com/cockroachdb/cockroach/commits/cee475331ca3629b503cd2e7c7919b72c98a5ca5):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/version-upgrade/run_1
	versionupgrade.go:258,versionupgrade.go:114,versionupgrade.go:189,versionupgrade.go:177,acceptance.go:69,acceptance.go:110,test_runner.go:760: write tcp 172.17.0.3:52698->35.192.202.190:26257: write: broken pipe

	cluster.go:1637,context.go:140,cluster.go:1626,test_runner.go:841: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2561359-1609916496-07-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 3: 5485
		4: 5484
		2: 5508
		1: dead
		Error: UNCLASSIFIED_PROBLEM: 1: dead
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1147
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:271
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:830
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1850
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:204
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (3) 1: dead
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError

```

<details><summary>More</summary><p>


Artifacts: [/acceptance/version-upgrade](https://teamcity.cockroachdb.com/viewLog.html?buildId=2561359&tab=artifacts#/acceptance/version-upgrade)
Related:
- #58329 roachtest: acceptance/version-upgrade failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-note-script-update](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-note-script-update) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #58307 roachtest: acceptance/version-upgrade failed  [A-schema-descriptors](https://api.github.com/repos/cockroachdb/cockroach/labels/A-schema-descriptors) [C-bug](https://api.github.com/repos/cockroachdb/cockroach/labels/C-bug) [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [branch-release-20.2](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.2)

- #58139 roachtest: acceptance/version-upgrade failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1)

- #53812 roachtest: acceptance/version-upgrade failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.2](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.2)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aacceptance%2Fversion-upgrade.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>


Jira issue: CRDB-3386

The first failure has the same stack trace as the mixed version test:

```
I210106 07:05:35.481781 1 util/log/flags.go:194  stderr capture started
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1d57f92]

goroutine 6945 [running]:
panic(0x4225c00, 0x7684030)
	/usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc000547430 sp=0xc000547368 pc=0x4e1b25
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:212
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:742 +0x413 fp=0xc000547460 sp=0xc000547430 pc=0x4f8873
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Version(0xc001899000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica.go:805 +0x72 fp=0xc0005474a0 sp=0xc000547460 pc=0x1d57f92
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).PurgeOutdatedReplicas.func1(0xc001899000, 0xc0013b7140)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:2814 +0x8c fp=0xc000547540 sp=0xc0005474a0 pc=0x1e226ac
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*storeReplicaVisitor).Visit(0xc0013b7140, 0xc000547600)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:396 +0x151 fp=0xc0005475a8 sp=0xc000547540 pc=0x1dd7211
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).VisitReplicas(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:2013
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).PurgeOutdatedReplicas(0xc000dca000, 0x55a3500, 0xc0013b7080, 0x200000014, 0xe00000000, 0xc001328f38, 0xc001ac43c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:2813 +0x1bf fp=0xc000547660 sp=0xc0005475a8 pc=0x1de32bf
github.com/cockroachdb/cockroach/pkg/server.(*migrationServer).PurgeOutdatedReplicas.func1(0xc000dca000, 0x4bad45, 0xc000547700)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/migration.go:184 +0x65 fp=0xc0005476a8 sp=0xc000547660 pc=0x3911065
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).VisitStores.func1(0x1, 0xc000dca000, 0xc000547700)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:131 +0x38 fp=0xc0005476d8 sp=0xc0005476a8 pc=0x1e26198
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*IntMap).Range(0xc000925840, 0xc000547790)
```

cc @irfansharif / @tbg

```I210209 18:39:37.489133 1 util/log/flags.go:195  [-] 1  stderr capture started
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1d58a72]

goroutine 2629 [running]:
panic(0x4328640, 0x7679090)
	/usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc0025074b8 sp=0xc0025073f0 pc=0x47d0a5
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:212
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:742 +0x413 fp=0xc0025074e8 sp=0xc0025074b8 pc=0x493df3
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Version(0xc001c96800, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica.go:814 +0x72 fp=0xc002507528 sp=0xc0025074e8 pc=0x1d58a72
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).PurgeOutdatedReplicas.func1(0xc001c96800, 0xc002a72d01)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:2722 +0x8c fp=0xc0025075c8 sp=0xc002507528 pc=0x1e2616c
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*storeReplicaVisitor).Visit(0xc002a72de0, 0xc002507688)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:375 +0x151 fp=0xc002507630 sp=0xc0025075c8 pc=0x1dda631
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).VisitReplicas(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:1988
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).PurgeOutdatedReplicas(0xc0014e4e00, 0x5716d60, 0xc002a72d50, 0x200000014, 0xe00000000, 0xc00107ba78, 0xc002027920)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:2721 +0x1bf fp=0xc0025076e8 sp=0xc002507630 pc=0x1de605f
github.com/cockroachdb/cockroach/pkg/server.(*migrationServer).PurgeOutdatedReplicas.func1(0xc0014e4e00, 0x4562c5, 0xc002507788)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/migration.go:184 +0x65 fp=0xc002507730 sp=0xc0025076e8 pc=0x39dbd05
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).VisitStores.func1(0x3, 0xc0014e4e00, 0xc002507788)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:131 +0x38 fp=0xc002507760 sp=0xc002507730 pc=0x1e2a158
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*IntMap).Range(0xc000ed1450, 0xc002507818)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/int_map.go:352 +0x130 fp=0xc0025077f8 sp=0xc002507760 pc=0x6cdbf0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).VisitStores(0xc000ed1420, 0xc002507888, 0x49f90b2, 0x17)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:130 +0x75 fp=0xc002507840 sp=0xc0025077f8 pc=0x1e08cf5
github.com/cockroachdb/cockroach/pkg/server.(*migrationServer).PurgeOutdatedReplicas(0xc00032f5c0, 0x5716d60, 0xc002a72c90, 0xc00216cd60, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/migration.go:183 +0x15e fp=0xc0025078c8 sp=0xc002507840 pc=0x399e77e
github.com/cockroachdb/cockroach/pkg/server/serverpb._Migration_PurgeOutdatedReplicas_Handler.func1(0x5716d60, 0xc002a72c90, 0x47df420, 0xc00216cd60, 0x0, 0x0, 0x0, 0xc0021f6a90)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/serverpb/migration.pb.go:487 +0x89 fp=0xc002507910 sp=0xc0025078c8 pc=0x268cc29
github.com/cockroachdb/cockroach/pkg/util/tracing.ServerInterceptor.func1(0x5716d60, 0xc002a72c90, 0x47df420, 0xc00216cd60, 0xc002a74200, 0xc002a74220, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/grpc_interceptor.go:126 +0x4bb fp=0xc002507a08 sp=0xc002507910 pc=0xe1b2fb
google.golang.org/grpc.getChainUnaryHandler.func1(0x5716d60, 0xc002a72c90, 0x47df420, 0xc00216cd60, 0xc000753ac8, 0xc93688, 0x46ab3c0, 0xc002a1b540)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0xe7 fp=0xc002507a88 sp=0xc002507a08 pc=0xca6647
github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1(0x5716d60, 0xc002a72c90, 0x47df420, 0xc00216cd60, 0xc002a74200, 0xc002a1b540, 0xc002a1b540, 0x20, 0x4413be0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:176 +0xa8 fp=0xc002507ad8 sp=0xc002507a88 pc=0x1948b68
google.golang.org/grpc.chainUnaryServerInterceptors.func1(0x5716d60, 0xc002a72c90, 0x47df420, 0xc00216cd60, 0xc002a74200, 0xc002a74220, 0xc000753ba0, 0x651ee6, 0x46c90c0, 0xc002a72c90)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:907 +0xd0 fp=0xc002507b40 sp=0xc002507ad8 pc=0xca64f0
github.com/cockroachdb/cockroach/pkg/server/serverpb._Migration_PurgeOutdatedReplicas_Handler(0x46bfea0, 0xc00032f5c0, 0x5716d60, 0xc002a72c90, 0xc002a23320, 0xc001a644c0, 0x5716d60, 0xc002a72c90, 0xc0023579d0, 0xa)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/serverpb/migration.pb.go:489 +0x150 fp=0xc002507bb0 sp=0xc002507b40 pc=0x25f07d0
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000fae1a0, 0x577f1c0, 0xc000c26900, 0xc0022b7e00, 0xc0005b15c0, 0x76a6748, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1082 +0x522 fp=0xc002507e40 sp=0xc002507bb0 pc=0xc93c62
google.golang.org/grpc.(*Server).handleStream(0xc000fae1a0, 0x577f1c0, 0xc000c26900, 0xc0022b7e00, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1405 +0xcc5 fp=0xc002507f68 sp=0xc002507e40 pc=0xc97d65
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000758190, 0xc000fae1a0, 0x577f1c0, 0xc000c26900, 0xc0022b7e00)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xa5 fp=0xc002507fb8 sp=0xc002507f68 pc=0xca61a5
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc002507fc0 sp=0xc002507fb8 pc=0x4b64e1
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:744 +0xa5
	```

Most recent failure is on my plate:

```
I210218 08:59:16.971744 1 util/log/flags.go:195  [-] 1  stderr capture started
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

goroutine 97 [running]:
panic(0x485d300, 0xc00310caa0)
	/usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc003310a50 sp=0xc003310988 pc=0x47d0a5
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0007b4580, 0x5900a60, 0xc000a564e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:233 +0x126 fp=0xc003310ab0 sp=0xc003310a50 pc=0x14c3866
runtime.call32(0x0, 0x51dae38, 0xc000d7ed68, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:540 +0x3e fp=0xc003310ae0 sp=0xc003310ab0 pc=0x4b4abe
runtime.reflectcallSave(0xc003310c20, 0x51dae38, 0xc000d7ed68, 0x18)
	/usr/local/go/src/runtime/panic.go:881 +0x58 fp=0xc003310b10 sp=0xc003310ae0 pc=0x47ca98
runtime.runOpenDeferFrame(0xc001250d80, 0xc000d7ed20, 0x0)
	/usr/local/go/src/runtime/panic.go:855 +0x2cd fp=0xc003310ba0 sp=0xc003310b10 pc=0x47c94d
panic(0x485d300, 0xc00310caa0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9 fp=0xc003310c68 sp=0xc003310ba0 pc=0x47cd19
runtime.goPanicIndex(0x0, 0x0)
	/usr/local/go/src/runtime/panic.go:88 +0xa5 fp=0xc003310cb0 sp=0xc003310c68 pc=0x47aba5
golang.org/x/net/trace.(*trace).addEvent(0xc0028a8b60, 0x43edb20, 0xc002ce2ba0, 0xc002c10000)
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:805 +0x71d fp=0xc003310d88 sp=0xc003310cb0 pc=0xc70ebd
golang.org/x/net/trace.(*trace).LazyPrintf(0xc0028a8b60, 0x4ae9480, 0x2, 0xc00313a250, 0x1, 0x1)
	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/net/trace/trace.go:834 +0x9a fp=0xc003310db8 sp=0xc003310d88 pc=0xc70ffa
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Recordf(0xc0028e3400, 0x4ae9480, 0x2, 0xc00313a230, 0x1, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:367 +0x1ba fp=0xc003310e60 sp=0xc003310db8 pc=0xe593fa
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Record(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:354
github.com/cockroachdb/cockroach/pkg/util/log.eventInternal(0xc0028e3400, 0x0, 0x0, 0x1, 0x1664cbe9f7f86f43, 0x61, 0x7832405, 0x1b, 0x41a, 0xc000b10f40, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:148 +0x778 fp=0xc003310f50 sp=0xc003310e60 pc=0xf68b18
github.com/cockroachdb/cockroach/pkg/util/log.Eventf(0x5920460, 0xc0028e3520, 0x4b76852, 0x24, 0xc003311190, 0x2, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:228 +0x158 fp=0xc003311080 sp=0xc003310f50 pc=0xf69558
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).refreshProposalsLocked(0xc000f09000, 0x5900a60, 0xc002c735f0, 0x0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:1050 +0x868 fp=0xc003311320 sp=0xc003311080 pc=0x1e03dc8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(0xc000f09000, 0x5900a60, 0xc002c735f0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:816 +0x1006 fp=0xc003311c30 sp=0xc003311320 pc=0x1e01446
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(0xc000f09000, 0x5900a60, 0xc002c735f0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:445 +0x110 fp=0xc003311d28 sp=0xc003311c30 pc=0x1e00330
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc0011b6a00, 0x5900a60, 0xc000a564e0, 0x129)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_raft.go:523 +0x134 fp=0xc003311ea0 sp=0xc003311d28 pc=0x1e56fd4
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker(0xc001bad040, 0x5900a60, 0xc000a564e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:284 +0x2c2 fp=0xc003311f20 sp=0xc003311ea0 pc=0x1e38de2
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker-fm(0x5900a60, 0xc000a564e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:229 +0x3e fp=0xc003311f48 sp=0xc003311f20 pc=0x1ea18be
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc0007b4580, 0x5900a60, 0xc000a564e0, 0xc000f2afc0, 0xc000b47b30)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:351 +0xb9 fp=0xc003311fb8 sp=0xc003311f48 pc=0x14c4cd9
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc003311fc0 sp=0xc003311fb8 pc=0x4b64e1
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:346 +0xfc
```

--- a/pkg/cmd/roachtest/tests/mixed_version_schemachange.go
+++ b/pkg/cmd/roachtest/tests/mixed_version_schemachange.go
@@ -27,7 +27,8 @@ func registerSchemaChangeMixedVersions(r registry.Registry) {
 		// This tests the work done for 20.1 that made schema changes jobs and in
 		// addition prevented making any new schema changes on a mixed cluster in
 		// order to prevent bugs during upgrades.
-		Cluster: r.MakeClusterSpec(4),
+		Cluster:    r.MakeClusterSpec(4),
+		NativeLibs: registry.LibGEOS,
 		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
 			maxOps := 100
 			concurrency := 5

--- a/pkg/cmd/roachtest/tests/mixed_version_schemachange.go
+++ b/pkg/cmd/roachtest/tests/mixed_version_schemachange.go
@@ -57,12 +57,6 @@ func runSchemaChangeWorkloadStep(loadNode, maxOps, concurrency int) versionStep 
 		t.L().Printf("Workload step run: %d", numFeatureRuns)
 		runCmd := []string{
 			"./workload run schemachange --verbose=1",
-			// The workload is still in development and occasionally discovers schema
-			// change errors so for now we don't fail on them but only on panics, server
-			// crashes, deadlocks, etc.
-			// TODO(spaskob): remove when https://github.com/cockroachdb/cockroach/issues/47430
-			// is closed.
-			"--tolerate-errors=true",
 			fmt.Sprintf("--max-ops %d", maxOps),
 			fmt.Sprintf("--concurrency %d", concurrency),
 			fmt.Sprintf("{pgurl:1-%d}", u.c.Spec().NodeCount),

--- a/pkg/workload/schemachange/operation_generator.go
+++ b/pkg/workload/schemachange/operation_generator.go
@@ -2373,6 +2373,18 @@ func (og *operationGenerator) insertRow(ctx context.Context, tx pgx.Tx) (stmt *o
 	if err != nil {
 		return nil, err
 	}
+	// If we aren't on 22.2 then disable the insert plugin, since 21.X
+	// can have schema instrospection queries fail due to an optimizer bug.
+	skipInserts, err := isClusterVersionLessThan(ctx, tx, clusterversion.ByKey(clusterversion.Start22_2))
+	if err != nil {
+		return nil, err
+	}
+	// If inserts are to be skipped, we will intentionally, target the insert towards
+	// a non-existent table, so that they become no-ops.
+	if skipInserts {
+		tableExists = false
+		tableName.SchemaName = "InvalidObjectName"
+	}
 	if !tableExists {
 		return makeOpStmtForSingleError(OpStmtDML,
 			fmt.Sprintf(

--- a/pkg/workload/schemachange/operation_generator.go
+++ b/pkg/workload/schemachange/operation_generator.go
@@ -1183,6 +1183,29 @@ func (og *operationGenerator) createTable(ctx context.Context, tx pgx.Tx) (*opSt
 	stmt := randgen.RandCreateTableWithColumnIndexNumberGenerator(og.params.rng, "table", tableIdx, databaseHasMultiRegion, og.newUniqueSeqNum)
 	stmt.Table = *tableName
 	stmt.IfNotExists = og.randIntn(2) == 0
+	trigramIsNotSupported, err := isClusterVersionLessThan(
+		ctx,
+		tx,
+		clusterversion.ByKey(clusterversion.TrigramInvertedIndexes))
+	if err != nil {
+		return nil, err
+	}
+	hasTrigramIdxUnsupported := func() bool {
+		if !trigramIsNotSupported {
+			return false
+		}
+		// Check if any of the indexes have trigrams involved.
+		for _, def := range stmt.Defs {
+			if idx, ok := def.(*tree.IndexTableDef); ok && idx.Inverted {
+				lastColumn := idx.Columns[len(idx.Columns)-1]
+				switch lastColumn.OpClass {
+				case "gin_trgm_ops", "gist_trgm_ops":
+					return true
+				}
+			}
+		}
+		return false
+	}()
 
 	tableExists, err := og.tableExists(ctx, tx, tableName)
 	if err != nil {
@@ -1197,6 +1220,11 @@ func (og *operationGenerator) createTable(ctx context.Context, tx pgx.Tx) (*opSt
 		{code: pgcode.DuplicateRelation, condition: tableExists && !stmt.IfNotExists},
 		{code: pgcode.UndefinedSchema, condition: !schemaExists},
 	}.add(opStmt.expectedExecErrors)
+	// Compatibility errors aren't guaranteed since the cluster version update is not
+	// fully transaction aware.
+	codesWithConditions{
+		{code: pgcode.FeatureNotSupported, condition: hasTrigramIdxUnsupported},
+	}.add(opStmt.potentialExecErrors)
 	opStmt.sql = tree.Serialize(stmt)
 	return opStmt, nil
 }

